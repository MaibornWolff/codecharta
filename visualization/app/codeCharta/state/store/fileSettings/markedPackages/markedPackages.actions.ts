import { CCAction, MarkedPackage } from "../../../../codeCharta.model"

export enum MarkedPackagesActions {
	SET_MARKED_PACKAGES = "SET_MARKED_PACKAGES",
	MARK_PACKAGE = "MARK_PACKAGE",
	UNMARK_PACKAGE = "UNMARK_PACKAGE"
}

export interface SetMarkedPackagesAction extends CCAction {
	type: MarkedPackagesActions.SET_MARKED_PACKAGES
	payload: MarkedPackage[]
}

export interface UnmarkPackageAction extends CCAction {
	type: MarkedPackagesActions.UNMARK_PACKAGE
	payload: number
}

export type MarkedPackagesAction = SetMarkedPackagesAction | UnmarkPackageAction

export function setMarkedPackages(markedPackages: MarkedPackage[] = defaultMarkedPackages): SetMarkedPackagesAction {
	return {
		type: MarkedPackagesActions.SET_MARKED_PACKAGES,
		payload: markedPackages
	}
}

export function unmarkPackage(index: number): UnmarkPackageAction {
	return {
		type: MarkedPackagesActions.UNMARK_PACKAGE,
		payload: index
	}
}

export const defaultMarkedPackages: MarkedPackage[] = []
