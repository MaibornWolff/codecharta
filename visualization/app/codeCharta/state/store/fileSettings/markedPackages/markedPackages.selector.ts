import { createSelector } from "../../../angular-redux/store"
import { fileSettingsSelector } from "../fileSettings.selector"

export const markedPackagesSelector = createSelector([fileSettingsSelector], fileSettings => fileSettings.markedPackages)
