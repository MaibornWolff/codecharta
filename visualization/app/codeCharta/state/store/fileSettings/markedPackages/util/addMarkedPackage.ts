import { MarkedPackage } from "../../../../../codeCharta.model"
import { getParent } from "../../../../../util/nodePathHelper"

export const addMarkedPackage = (markedPackagesMap: Map<string, MarkedPackage>, { path, color }: MarkedPackage) => {
	const directMarkedParentPackage = getParent(markedPackagesMap, path)

	if (!directMarkedParentPackage || directMarkedParentPackage.color !== color) {
		markedPackagesMap.set(path, {
			path,
			color
		})
	}

	for (const [key, markedPackage] of markedPackagesMap) {
		if (markedPackage.path === path) {
			if (markedPackage.color !== color) {
				markedPackagesMap.delete(key)
			}
		} else if (markedPackage.path.startsWith(path)) {
			// Remove marked packages with color identical to their parent marked package
			const markedPackageParent = getParent(markedPackagesMap, markedPackage.path)
			if (markedPackageParent && markedPackageParent.color === markedPackage.color) {
				markedPackagesMap.delete(key)
			}
		}
	}
}
