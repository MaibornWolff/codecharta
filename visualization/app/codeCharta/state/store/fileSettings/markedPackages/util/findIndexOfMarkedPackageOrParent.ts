import { MarkedPackage } from "../../../../../codeCharta.model"

/** @return index of exact match if there is an exact match. Otherwise return index of parent or -1  */
export const findIndexOfMarkedPackageOrParent = (markedPackages: MarkedPackage[], nodePath: string) => {
	const indexOfNodePath = markedPackages.findIndex(mp => mp.path === nodePath)
	if (indexOfNodePath !== -1) return indexOfNodePath

	let indexOfNearestParent = -1
	for (let loopIndex = 0; loopIndex < markedPackages.length; loopIndex++) {
		const markedPackage = markedPackages[loopIndex]
		if (
			nodePath.startsWith(markedPackage.path) &&
			(indexOfNearestParent === -1 || markedPackages[indexOfNearestParent].path.length < markedPackage.path.length)
		)
			indexOfNearestParent = loopIndex
	}

	return indexOfNearestParent
}
