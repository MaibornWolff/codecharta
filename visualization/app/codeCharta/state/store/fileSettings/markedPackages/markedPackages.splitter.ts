import { MarkedPackagesAction, setMarkedPackages } from "./markedPackages.actions"
import { MarkedPackage } from "../../../../codeCharta.model"

export function splitMarkedPackagesAction(payload: MarkedPackage[]): MarkedPackagesAction {
	return setMarkedPackages(payload)
}
