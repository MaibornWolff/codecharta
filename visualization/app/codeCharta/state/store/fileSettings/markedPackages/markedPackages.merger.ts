import { getUpdatedBlacklistItemPath } from "../../../../util/nodePathHelper"
import { CCFile, MarkedPackage } from "../../../../codeCharta.model"

export function getMergedMarkedPackages(inputFiles: CCFile[], withUpdatedPath: boolean) {
	const markedPackages: Map<string, MarkedPackage> = new Map()

	if (inputFiles.length === 1) {
		return inputFiles[0].settings.fileSettings.markedPackages
	}

	for (const inputFile of inputFiles) {
		if (inputFile.settings.fileSettings.markedPackages) {
			for (const oldMarkedPackages of inputFile.settings.fileSettings.markedPackages) {
				const markedPackage: MarkedPackage = {
					path: withUpdatedPath
						? getUpdatedBlacklistItemPath(inputFile.fileMeta.fileName, oldMarkedPackages.path)
						: oldMarkedPackages.path,
					color: oldMarkedPackages.color
				}
				markedPackages.set(`${markedPackage.path}|${markedPackage.color}`, markedPackage)
			}
		}
	}
	return [...markedPackages.values()]
}
