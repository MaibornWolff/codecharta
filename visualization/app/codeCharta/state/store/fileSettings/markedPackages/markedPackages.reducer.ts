import { MarkedPackagesAction, MarkedPackagesActions, setMarkedPackages } from "./markedPackages.actions"
import { removeEntryAtIndexFromArray } from "../../../../util/reduxHelper"

export function markedPackages(state = setMarkedPackages().payload, action: MarkedPackagesAction) {
	switch (action.type) {
		case MarkedPackagesActions.SET_MARKED_PACKAGES:
			return action.payload
		case MarkedPackagesActions.UNMARK_PACKAGE:
			return removeEntryAtIndexFromArray(state, action.payload)
		default:
			return state
	}
}
