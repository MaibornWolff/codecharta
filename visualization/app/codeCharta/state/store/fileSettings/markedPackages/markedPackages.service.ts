import { StoreService, StoreSubscriber } from "../../../store.service"
import { IRootScopeService } from "angular"
import { MarkedPackagesActions, setMarkedPackages } from "./markedPackages.actions"
import { MarkedPackage } from "../../../../codeCharta.model"
import { getMergedMarkedPackages } from "./markedPackages.merger"
import { FilesService, FilesSelectionSubscriber } from "../../files/files.service"
import { Files } from "../../../../model/files"
import { isActionOfType } from "../../../../util/reduxHelper"

export interface MarkedPackagesSubscriber {
	onMarkedPackagesChanged(markedPackages: MarkedPackage[])
}

export class MarkedPackagesService implements StoreSubscriber, FilesSelectionSubscriber {
	private static MARKED_PACKAGES_CHANGED_EVENT = "marked-packages-changed"

	constructor(private $rootScope: IRootScopeService, private storeService: StoreService) {
		StoreService.subscribe(this.$rootScope, this)
		FilesService.subscribe(this.$rootScope, this)
	}

	public onStoreChanged(actionType: string) {
		if (isActionOfType(actionType, MarkedPackagesActions)) {
			this.notify(this.select())
		}
	}

	public onFilesSelectionChanged(fileStates: Files) {
		this.merge(fileStates)
	}

	private merge(files: Files) {
		const visibleFiles = files.getVisibleFiles()
		const withUpdatedPath = files.isPartialState()
		const newMarkedPackages = getMergedMarkedPackages(visibleFiles, withUpdatedPath)
		this.storeService.dispatch(setMarkedPackages(newMarkedPackages))
	}

	private select() {
		return this.storeService.getState().fileSettings.markedPackages
	}

	private notify(newState: MarkedPackage[]) {
		this.$rootScope.$broadcast(MarkedPackagesService.MARKED_PACKAGES_CHANGED_EVENT, { markedPackages: newState })
	}

	public static subscribe($rootScope: IRootScopeService, subscriber: MarkedPackagesSubscriber) {
		$rootScope.$on(MarkedPackagesService.MARKED_PACKAGES_CHANGED_EVENT, (event, data) => {
			subscriber.onMarkedPackagesChanged(data.markedPackages)
		})
	}
}
