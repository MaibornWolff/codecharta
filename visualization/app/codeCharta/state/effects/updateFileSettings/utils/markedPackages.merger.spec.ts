import { TEST_FILE_DATA } from "../../../../util/dataMocks"
import { getMergedMarkedPackages } from "./markedPackages.merger"
import { CCFile, MarkedPackage } from "../../../../codeCharta.model"
import { clone } from "../../../../util/clone"

describe("MarkedPackagesMerger", () => {
	describe("getMergedMarkedPackages", () => {
		let mp1: MarkedPackage
		let mp2: MarkedPackage
		let mp3: MarkedPackage
		let mp4: MarkedPackage

		let file1: CCFile
		let file2: CCFile

		beforeEach(() => {
			file1 = clone(TEST_FILE_DATA)
			file1.fileMeta.fileName = "file1"

			file2 = clone(TEST_FILE_DATA)
			file2.fileMeta.fileName = "file2"

			mp1 = {
				path: "/root/nodeA",
				color: "#ABABAB"
			}

			mp2 = {
				path: "/root/nodeB",
				color: "#FFFFFF"
			}

			mp3 = {
				path: "/root/nodeA",
				color: "#ABABAB"
			}

			mp4 = {
				path: "/root/nodeA",
				color: "#ABABAB"
			}
		})

		it("should merge empty markedPackage-arrays", () => {
			file1.settings.fileSettings.markedPackages = []
			file2.settings.fileSettings.markedPackages = []
			expect(getMergedMarkedPackages([file1, file2], false)).toEqual([])
		})

		it("should merge different markedPackages", () => {
			file1.settings.fileSettings.markedPackages = [mp1]
			file2.settings.fileSettings.markedPackages = [mp2]
			expect(getMergedMarkedPackages([file1, file2], false)).toMatchSnapshot()
		})

		it("should merge all markedPackages if one file does not contain markedPackages", () => {
			file1.settings.fileSettings.markedPackages = [mp1, mp2]
			file2.settings.fileSettings.markedPackages = null
			expect(getMergedMarkedPackages([file1, file2], false)).toMatchSnapshot()
		})

		it("should merge markedPackage-attributes for the same markedPackage paths", () => {
			file1.settings.fileSettings.markedPackages = [mp1]
			file2.settings.fileSettings.markedPackages = [mp3]
			expect(getMergedMarkedPackages([file1, file2], false)).toMatchSnapshot()
		})

		it("should overwrite duplicated markedPackage-attributes for the same markedPackage", () => {
			file1.settings.fileSettings.markedPackages = [mp1]
			file2.settings.fileSettings.markedPackages = [mp4]
			expect(getMergedMarkedPackages([file1, file2], false)).toMatchSnapshot()
		})
	})
})
