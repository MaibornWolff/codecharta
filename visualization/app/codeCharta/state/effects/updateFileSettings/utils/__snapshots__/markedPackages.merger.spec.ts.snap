// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarkedPackagesMerger getMergedMarkedPackages should merge all markedPackages if one file does not contain markedPackages 1`] = `
[
  {
    "color": "#ABABAB",
    "path": "/root/nodeA",
  },
  {
    "color": "#FFFFFF",
    "path": "/root/nodeB",
  },
]
`;

exports[`MarkedPackagesMerger getMergedMarkedPackages should merge different markedPackages 1`] = `
[
  {
    "color": "#ABABAB",
    "path": "/root/nodeA",
  },
  {
    "color": "#FFFFFF",
    "path": "/root/nodeB",
  },
]
`;

exports[`MarkedPackagesMerger getMergedMarkedPackages should merge markedPackage-attributes for the same markedPackage paths 1`] = `
[
  {
    "color": "#ABABAB",
    "path": "/root/nodeA",
  },
]
`;

exports[`MarkedPackagesMerger getMergedMarkedPackages should overwrite duplicated markedPackage-attributes for the same markedPackage 1`] = `
[
  {
    "color": "#ABABAB",
    "path": "/root/nodeA",
  },
]
`;
